### PR TITLE
refactor visited list object

### PR DESCRIPTION
**Pull request by alonre24**
Refactor of this object - add some changes to the API, logic remains the same

### CODE REVIEW

File: src/VecSim/algorithms/hnsw/visited_nodes_handler.h [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+typedef ushort tag_t;
+
+/**
+ * Used as a singleton that is responsible for marking nodes that were visited in the graph scan.
+ * Every scan has a "pseudo unique" tag which is associated with this specific scan, and nodes
+ * that were visited in this particular scan are tagged with this tag. The tags range from
+ * 1-MAX_USHORT, and we reset the tags after we complete MAX_USHORT scans.
+ */
+class VisitedNodesHandler : public VecsimBaseObject {
```
Comment (alonre24) [inline src/VecSim/algorithms/hnsw/visited_nodes_handler.h:18]: add `private`

File: src/VecSim/algorithms/hnsw/visited_nodes_handler.h [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    VisitedNodesHandler(uint cap, const std::shared_ptr<VecSimAllocator> &allocator)
+        : VecsimBaseObject(allocator) {
+        cur_tag = 0;
+        num_elements = cap;
+        elements_tags =
+            reinterpret_cast<tag_t *>(allocator->allocate(sizeof(tag_t) * num_elements));
+        memset(elements_tags, 0, sizeof(tag_t) * num_elements);
+    }
+
+    tag_t getFreshTag() {
```
Comment (alonre24) [inline src/VecSim/algorithms/hnsw/visited_nodes_handler.h:33]: add doc

File: src/VecSim/algorithms/hnsw/visited_nodes_handler.h [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+            memset(elements_tags, 0, sizeof(tag_t) * num_elements);
+            cur_tag++;
+        }
+        return cur_tag;
+    }
+
+    void visitNode(uint node_id, tag_t tag) { elements_tags[node_id] = tag; }
+
+    tag_t getNodeTag(uint node_id) { return elements_tags[node_id]; }
+
```
Comment (alonre24) [inline src/VecSim/algorithms/hnsw/visited_nodes_handler.h:45]: inline

File: src/VecSim/algorithms/hnsw/visited_nodes_handler.h [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+public:
+    VisitedNodesHandler(uint cap, const std::shared_ptr<VecSimAllocator> &allocator);
+
+    // Return unused tag for marking the visited nodes. The tags are cyclic, so whenever we reach
+    // zero, we reset the tags of all the nodes (and use 1 as the fresh tag)
+    tag_t getFreshTag();
+
+    inline void visitNode(uint node_id, tag_t tag) { elements_tags[node_id] = tag; }
+    inline tag_t getNodeTag(uint node_id) { return elements_tags[node_id]; }
```
Comment (DvirDukhan) [inline src/VecSim/algorithms/hnsw/visited_nodes_handler.h:32]: can you add docs for `visitNode` and `getNodeTag` please?

File: src/VecSim/algorithms/hnsw/visited_nodes_handler.cpp [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+VisitedNodesHandlerPool::~VisitedNodesHandlerPool() {
+    while (!pool.empty()) {
+        VisitedNodesHandler *handler = pool.front();
+        pool.pop_front();
+        delete handler;
+    }
+}
+
+} // namespace hnswlib
```
Comment (DvirDukhan) [inline src/VecSim/algorithms/hnsw/visited_nodes_handler.cpp:59]: empty line

**Pull request merged**